i8kctl.1
.TH i8kctl 1 2017-02-26 "Vitor Augusto"
.SH "NAME"
.LP
i8kctl \- Utility to get information, and set fans states
.SH "SYNOPSIS"
\fBi8kctl\fP [\fI<param>\fP [\fI<args>\fP]]
.SH "DESCRIPTION"
.LP
The i8kctl can be used to retrieve the information listed below. It does that
by means of querying the system through the dell-smm-hwmon kernel module. When invoked
without arguments it reports all the information at once in one line of text.
The information is the same as the file \fB/proc/i8k\fR.
.IP
1.  i8k output format version
.br
2.  Dell bios version
.br
3.  service tag (later 'machine id')
.br
4.  cpu temperature
.br
5.  left fan status
.br
6.  right fan status
.br
7.  left fan speed
.br
8.  right fan speed
.br
9.  ac power status
.br
10. fn buttons status
.LP
The following parameters select one of the
information: \fBversion\fR, \fBbios\fR, \fBid\fR,
\fBtemp\fR, \fBfan\fR, \fBspeed\fR, \fBac\fR and \fBfn\fR.
.LP
Some examples of usage and output are below.
.IP
$ i8kctl
.br
1.0 A17 B5W123K 52 2 1 8040 6420 1 2
.IP
$ i8kctl temp
.br
52
.IP
$ i8kctl fan
.br
2 1
.LP
The \fBfan\fR parameter accepts two optional arguments for the fans
states. The state arguments can be:
.IP
0  turn the fan off
.br
1  set low speed
.br
2  set high speed
.br
\-  don't change the state of this fan
.LP
And the following is an example of usage:
.IP
$ i8kctl fan \- 2
.LP
This sets the right fan to high speed and leaves the left unchanged.
It should be noted that if the \fBi8kmon(1)\fR daemon is used to
control the fans, setting the speed with \fBi8kctl fan\fP is pointless since the
daemon override regularly the speed with its own value.
.LP
If \fBi8kctl fan\fR is invoked without the optional arguments, the states
of the fans are returned, and the output may contains a -1 which indicates that the
corresponding fan does not exist.
.SH "FILES"
.LP
\fI/proc/i8k\fP
.SH "AUTHORS"
.LP
2013-2017 Vitor Augusto <vitorafsr@gmail.com>
.br
2001-2009 Massimo Dal Zotto <dz@debian.org>
.SH "COPYRIGHT"
.LP
i8kctl is part of the i8kutils package. All files are distributed under the GNU
General Public License (GPL). On Debian GNU/Linux systems, the complete text of
the GNU General Public License can be found in `/usr/share/common-licenses/GPL'.
.SH "SEE ALSO"
.LP
i8kmon(1)


i8kctl.c
/*
 * i8kctl.c -- Utility to query the i8k kernel module on Dell laptops to
 * retrieve information
 *
 * Copyright (C) 2013-2017 Vitor Augusto <vitorafsr@gmail.com>
 * Copyright (C) 2001  Massimo Dal Zotto <dz@debian.org>
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2, or (at your option) any
 * later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 */

#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <stdlib.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/ioctl.h>
#include <unistd.h>

#include "i8k.h"
#include "i8kctl.h"

static int i8k_fd;

char *
i8k_get_bios_version()
{
    char i8k_ioctl_bios_version[4];

    char i8k_proc_format[16];
    char i8k_proc_bios_version[4];
    char i8k_proc_serial_version[16];
    int i8k_proc_cpu_temp;
    int i8k_proc_left_fan;
    int i8k_proc_right_fan;
    int i8k_proc_left_speed;
    int i8k_proc_right_speed;
    int i8k_proc_ac_power;
    int i8k_proc_fn_key;

    char proc_i8k_str[64];
    int args[1];
    int rc_read;
    int ret_nargs;

    if ((rc_read=read(i8k_fd, proc_i8k_str, 64*sizeof(char))) != -1) {

        ret_nargs = sscanf(proc_i8k_str, "%s %s %s %d %d %d %d %d %d %d\n",
               i8k_proc_format,
               i8k_proc_bios_version,
               i8k_proc_serial_version,
               &i8k_proc_cpu_temp,
               &i8k_proc_left_fan,
               &i8k_proc_right_fan,
               &i8k_proc_left_speed,
               &i8k_proc_right_speed,
               &i8k_proc_ac_power,
               &i8k_proc_fn_key);

        if (ret_nargs == 10) {
            if (strncmp(i8k_proc_format, I8K_PROC_FMT, 16) == 0) {

                // Stop this function here and return.
                // Information from /proc/i8k is more complete.
                // So it is preferred here.
                // ioctl for BIOS version at the kernel module i8k returns
                // only the SMM version. The information of bios version in
                // '/proc/i8k' is from SMM data and DMI table.
                return strdup(i8k_proc_bios_version);
            }

        }
    }

    // Useless in some Dell systems for i8k kernel module dated 2013-05-30.
    if (ioctl(i8k_fd, I8K_BIOS_VERSION, &args) != 0) {
        i8k_ioctl_bios_version[0] = (args[0] >> 16) & 0xff;
        i8k_ioctl_bios_version[1] = (args[0] >>  8) & 0xff;
        i8k_ioctl_bios_version[2] = (args[0]      ) & 0xff;
        i8k_ioctl_bios_version[3] = '\0';
        return strdup(i8k_ioctl_bios_version);
    }

    return 0;
}

char *
i8k_get_machine_id()
{
    char args[16];
    int rc;

    if ((rc=ioctl(i8k_fd, I8K_MACHINE_ID, &args)) < 0) {
	return NULL;
    }

    return strdup(args);
}

int
i8k_set_fan(int fan, int speed)
{
    int args[2];
    int rc;

    args[0] = fan;
    args[1] = speed;
    if ((rc=ioctl(i8k_fd, I8K_SET_FAN, &args)) < 0) {
	return rc;
    }

    return args[0];
}

int
i8k_get_fan_status(int fan)
{
    int args[1];
    int rc;

    args[0] = fan;
    if ((rc=ioctl(i8k_fd, I8K_GET_FAN, &args)) < 0) {
	return rc;
    }

    return args[0];
}

int
i8k_get_fan_speed(int fan)
{
    int args[1];
    int rc;

    args[0] = fan;
    if ((rc=ioctl(i8k_fd, I8K_GET_SPEED, &args)) < 0) {
	return rc;
    }

    return args[0];
}

int
i8k_get_cpu_temp()
{
    int args[1];
    int rc;

    if ((rc=ioctl(i8k_fd, I8K_GET_TEMP, &args)) < 0) {
	return rc;
    }

    return args[0];
}

int
i8k_get_power_status()
{
    int args[1];
    int rc;

    if ((rc=ioctl(i8k_fd, I8K_POWER_STATUS, &args)) < 0) {
	return rc;
    }

    return args[0];
}

int
i8k_get_fn_status()
{
    int args[1];
    int rc;

    if ((rc=ioctl(i8k_fd, I8K_FN_STATUS, &args)) < 0) {
	return rc;
    }

    return args[0];
}

int
fan(int argc, char **argv)
{
    int left, right;

    if ((argc > 1) && isdigit(argv[1][0])) {
	left = i8k_set_fan(I8K_FAN_LEFT, atoi(argv[1]));
    } else {
	left = i8k_get_fan_status(I8K_FAN_LEFT);
    }

    if ((argc > 2) && isdigit(argv[2][0])) {
	right = i8k_set_fan(I8K_FAN_RIGHT, atoi(argv[2]));
    } else {
	right = i8k_get_fan_status(I8K_FAN_RIGHT);
    }

    printf("%d %d\n", left, right);
    return 0;
}

int
fan_speed(int argc, char **argv)
{
    int left, right;

    left = i8k_get_fan_speed(I8K_FAN_LEFT);
    right = i8k_get_fan_speed(I8K_FAN_RIGHT);

    printf("%d %d\n", left, right);
    return 0;
}

int
bios_version()
{
    char *version;

    if ((version=i8k_get_bios_version()) == NULL) {
	version = "?";
    }

    printf("%s\n", version);
    return 0;
}

int
machine_id()
{
    char *machine_id;

    if ((machine_id=i8k_get_machine_id()) == NULL) {
	machine_id = "?";
    }

    printf("%s\n", machine_id);
    return 0;
}

int
cpu_temperature()
{
    printf("%d\n", i8k_get_cpu_temp());
    return 0;
}

int
ac_status()
{
    printf("%d\n", i8k_get_power_status());
    return 0;
}

int
fn_key()
{
    printf("%d\n", i8k_get_fn_status());
    return 0;
}

int
status()
{
    int fn_key, cpu_temp, ac_power;
    int left_fan, right_fan, left_speed, right_speed;
    char *bios_version, *bios_machine_id;

    bios_version    = i8k_get_bios_version();
    bios_machine_id = i8k_get_machine_id();
    cpu_temp        = i8k_get_cpu_temp();
    left_fan        = i8k_get_fan_status(I8K_FAN_LEFT);
    right_fan       = i8k_get_fan_status(I8K_FAN_RIGHT);
    left_speed      = i8k_get_fan_speed(I8K_FAN_LEFT);
    right_speed     = i8k_get_fan_speed(I8K_FAN_RIGHT);
    ac_power        = i8k_get_power_status();
    fn_key          = i8k_get_fn_status();

    /*
     * Info:
     *
     * 1)  Proc format version (this will change if format changes)
     * 2)  BIOS version
     * 3)  BIOS machine ID
     * 4)  Cpu temperature
     * 5)  Left fan status
     * 6)  Right fan status
     * 7)  Left fan speed
     * 8)  Right fan speed
     * 9)  AC power
     * 10) Fn Key status
     */
    printf("%s %s %s %d %d %d %d %d %d %d\n",
	   I8K_PROC_FMT,
	   bios_version,
	   bios_machine_id,
	   cpu_temp,
	   left_fan,
	   right_fan,
	   left_speed,
	   right_speed,
	   ac_power,
	   fn_key);

    return 0;
}

void
usage()
{
    printf("Usage: i8kctl [fan [<l> <r>] | speed | version | bios | id | temp" \
           "| ac | fn]\n");
    printf("       i8kctl [-h]\n");
}

#ifdef LIB
void init()
{
    i8k_fd = open(I8K_PROC, O_RDONLY);
    if (i8k_fd < 0)
    {
        perror("can't open " I8K_PROC);
        exit(-1);
    }
}
void finish()
{
    close(i8k_fd);
}
#else
int
main(int argc, char **argv)
{
    if (argc >= 2) {
	if ((strcmp(argv[1],"-h")==0) || (strcmp(argv[1],"--help")==0)) {
	    usage();
	    exit(0);
	}
    }

    i8k_fd = open(I8K_PROC, O_RDONLY);
    if (i8k_fd < 0) {
        perror("can't open " I8K_PROC);
        exit(-1);
    }

    /* -2 as a magic number: if var 'ret' reachs the end of main() as -2, than
     * no command was executed, and the user input was an invalid command
     */
    int ret = -2;

    /* No args, print status: same output as 'cat /proc/i8k' */
    if (argc < 2) {
        ret = status();
        close(i8k_fd);
        return ret;
    }

    if (strcmp(argv[1],"fan")==0) {
        argc--; argv++;
        ret = fan(argc,argv);
    }
    else if (strcmp(argv[1],"version")==0) {
        printf("%s\n", I8K_PROC_FMT);
        ret = 0;
    }
    else if (strcmp(argv[1],"speed")==0) {
        ret = fan_speed(argc,argv);
    }
    else if (strcmp(argv[1],"bios")==0) {
        ret = bios_version();
    }
    else if (strcmp(argv[1],"id")==0) {
        ret = machine_id();
    }
    else if (strcmp(argv[1],"temp")==0) {
        ret = cpu_temperature();
    }
    else if (strcmp(argv[1],"ac")==0) {
        ret = ac_status();
    }
    else if (strcmp(argv[1],"fn")==0) {
        ret = fn_key();
    }

    close(i8k_fd);

    if (ret == -2) // no command executed
        fprintf(stderr,"invalid arg: %s\n", argv[1]);

    return 0;
}
#endif
